File: pset1/cash/cash.c
// CS50x Problem Set 1 — Cash (greedy algorithm)
//
// Prompts for an amount of change owed (in dollars) and prints the minimum number of
// coins (quarters, dimes, nickels, pennies) needed to make that change, using a greedy
// algorithm. Reprompts until the user enters a non-negative dollar amount.

#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    // Prompt for change in dollars; reject negatives.
    float dollars;
    do
    {
        dollars = get_float("Change owed: ");
    }
    while (dollars < 0);

    // Convert to cents, rounding to the nearest integer to avoid float error.
    int cents = round(dollars * 100);

    int coins = 0;

    // Greedily use the largest coin that fits, repeatedly.
    coins += cents / 25;
    cents %= 25;

    coins += cents / 10;
    cents %= 10;

    coins += cents / 5;
    cents %= 5;

    coins += cents; // remaining pennies

    printf("%i\n", coins);
}

File: pset6/cash/cash.py
# CS50x Problem Set 6 — Cash (greedy, Python)
#
# Prompts for change owed (in dollars) and prints the minimum number of coins needed to
# make that amount using quarters, dimes, nickels, and pennies (a greedy algorithm).


def get_cents():
    while True:
        try:
            dollars = float(input("Change owed: "))
        except ValueError:
            continue
        if dollars >= 0:
            # Round to the nearest cent to avoid floating-point error.
            return round(dollars * 100)


def main():
    cents = get_cents()
    coins = 0
    for value in (25, 10, 5, 1):
        coins += cents // value
        cents %= value
    print(coins)


if __name__ == "__main__":
    main()
